src/main/java/org/openstreetmap/atlas/geography/sharding/GeoHashSharding.java

```diff
@@ -0,0 +1,57 @@
+package org.openstreetmap.atlas.geography.sharding;
+
+import org.openstreetmap.atlas.exception.CoreException;
+import org.openstreetmap.atlas.geography.GeometricSurface;
+import org.openstreetmap.atlas.geography.Location;
+import org.openstreetmap.atlas.geography.PolyLine;
+import org.openstreetmap.atlas.utilities.collections.Iterables;
+
+/**
+ * @author matthieun
+ */
+public class GeoHashSharding implements Sharding
+{
+    private static final long serialVersionUID = -7355746343440111174L;
+
+    private final int precision;
+
+    public GeoHashSharding(final int precision)
+    {
+        GeoHashTile.validatePrecision(precision);
+        this.precision = precision;
+    }
+
+    @Override
+    public Iterable<Shard> neighbors(final Shard shard)
+    {
+        if (shard instanceof GeoHashTile)
+        {
+            return ((GeoHashTile) shard).neighbors();
+        }
+        else
+        {
+            throw new CoreException("Cannot have neighbors from another type of shard.");
```

Comment:
What about:
```java
throw new CoreException("shard parameter was of invalid type {}", shard.getClass().getName());
```
This may be slightly more helpful as an error message.